clients/roscpp/src/libros/service_server_link.cpp

```diff
@@ -357,15 +358,30 @@ bool ServiceServerLink::call(const SerializedMessage& req, SerializedMessage& re
     processNextCall();
   }
 
+  bool interrupted = false;
   {
     boost::mutex::scoped_lock lock(info->finished_mutex_);
 
     while (!info->finished_)
     {
-      info->finished_condition_.wait(lock);
+    	if(timeout > 0)
+    	{
+    		boost::chrono::milliseconds duration(static_cast<int>(timeout * 1000));
+    		if(info->finished_condition_.wait_for(lock, duration))
+    		{
+    			ROS_ERROR("Service [%s] call failed: no response for %fsec", service_name_.c_str(), timeout);
+    			interrupted = true;
+    			break;
+    		}
+    	}
+    	else
+    		info->finished_condition_.wait(lock);
     }
   }
 
+  if(interrupted)
+	  this->clearCalls();
```

Comment:
This function call seems to require the same lock as used above: `info->finished_mutex_`

Reply:
clearCalls is mutex-protected by itself. Protecting it by info->finished_mutex_ will cause a deadlock at ServiceServerLink::cancelCall